File: src/main/java/main/FtpFilesToHdfs.java
package main;

import job.demo.GlobalInstance;
import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;
import org.apache.commons.net.ftp.FTPFile;
import org.apache.hadoop.fs.FSDataOutputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IOUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/*@Deprecated不推荐使用 修改为scala版本*/
@Deprecated
public class FtpFilesToHdfs {

    private static String hdfsMasterBase = "hdfs://";
    //"hdfs://"//
    private static String hdfsMasterPath = "hdfs://";
    private static String ftpIP = "";
    private static String ftpUser = "root";
    private static String ftpPW = "root123";
    /*hdfs目录*/
    private static String hdfsWorkPath = "/root";
    /*ftp目录*/
    private static String ftpBasePath = "/home/doc";
    /*hdfs用户名*/
    private static String hdfsUserName = "root";
    /*ftp文件reg*/
    private static String regExp = "^[A-Za-z]{3}_(SEGMENT|CLASS){1}_\\d{8}\\.csv$";
    /*ftp目录搜索深度*/
    private static int dept = 3;
    private boolean NULLABLE = true;

    public static void main(String[] args) {
        if(args != null){
            System.out.println("args length"+args.length);
            for (String s:args){
                System.out.println("args:"+s);
            }
        }
        if(args != null && args.length >4){
            ftpBasePath = args[0];
            hdfsWorkPath = args[1];
            hdfsUserName = args[2];
            regExp = args[3];
            dept = Integer.valueOf(args[4]);
        }
        SparkConf conf = new SparkConf().setAppName("ftp files download")
                .set("spark.sql.parquet.compression.codec","uncomoressed");
        //gzip
        SparkContext spark = new SparkContext(conf);
        System.setProperty("HADOOP_USER_NAME",hdfsUserName);
        try{
            System.out.println("==start==");
            ftpFilesDownload(hdfsWorkPath,ftpBasePath,regExp,dept);
            System.out.println("end");
        }catch (Exception e){
            e.printStackTrace();
        }
        spark.stop();
    }


    private static void ftpFilesDownload(
            String hdfsPath,
            String ftpPath,
            String regExp,
            int dept
    )throws Exception{
        List<String> files = new ArrayList<>(500);
        final FTPClient[] ftp = {ftpReconnect()};
        ftpListFile(ftp[0], ftpPath, regExp, files, dept);
        ftp[0].disconnect();
        for (String file:files){
            try{
                int separatorIndex = file.lastIndexOf("/");
                String fileName = file.substring(separatorIndex+1);
                String filePath = file.substring(0,separatorIndex);
                ftp[0] = ftpReconnect();
                ftp[0].changeWorkingDirectory(filePath);
                InputStream inputStream = ftp[0].retrieveFileStream(fileName);
                FileSystem fileSystem = GlobalInstanceJava.getHdfsFileSystem(hdfsMasterBase,hdfsUserName);
                FSDataOutputStream outputStream = fileSystem.create(new Path(hdfsPath+fileName),true);
                IOUtils.copyBytes(inputStream,outputStream,81920);
                System.out.println(file+"downloaded");
                outputStream.close();
                inputStream.close();
                ftp[0].disconnect();
            }catch (Exception e){
                e.printStackTrace();
            }
        }

    }
    //递归遍历fatp文件
    private static void ftpListFile(FTPClient ftpClient,
                                    String ftpPath,
                                    String regExp,
                                    List<String> files,
                                    int dept) throws Exception{
        int anotherDept = dept;
        if (anotherDept <= 0){
            return;
        }
        FTPClient ftp = ftpClient;
        if(!ftp.isConnected()){
            ftp = ftpReconnect();
        }
        String directory = ftpPath;
        //更换目录到当前目录
        ftp.changeWorkingDirectory(directory);
        FTPFile[] ftpFiles = ftp.listFiles();
        anotherDept--;
        for (FTPFile file:ftpFiles) {
            if (file.isFile()){
                if(file.getName().matches(regExp)){
                    files.add(directory+file.getName());
                    System.out.println("file:"+directory+file.getName()+",size"+file.getSize());
                }
            }else {
                System.out.println("dir:"+directory+file.getName()+",dept:"+dept);
                ftpListFile(ftp,directory+file.getName()+"/",regExp,files,anotherDept);
            }

        }

    }

    private static FTPClient ftpReconnect() throws Exception{
        FTPClient ftp = new FTPClient();
        ftp.setControlEncoding("UTF-8");
        ftp.setDefaultTimeout(72000);
        ftp.connect(ftpIP,21);
        ftp.login(ftpUser,ftpPW);
        ftp.enterLocalPassiveMode();
        return ftp;
    }


    private static FTPClient ftpReconnect(String ip,int port,String user,String password) throws Exception{
        FTPClient ftp = new FTPClient();
        ftp.connect(ip,port);
        ftp.login(user,password);
        return ftp;
    }


}

File: src/main/scala/job/demo/FtpFilesToHdfs.scala
package job.demo

import java.io._

import org.apache.commons.net.ftp
import org.apache.commons.net.ftp.{FTP, FTPClient}
import org.apache.hadoop.fs.Path
import org.apache.hadoop.io.IOUtils
import org.apache.spark.{SparkConf, SparkContext}

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.io.Source
/*ftp下载文件到hdfs*/
object FtpFilesToHdfs {

    var NULLABLE = true
    //生产环境hdfs主备uri
    val hdfsUris = Array("hdfs://", "hdfs://")
    var hdfsMasterBase = "hdfs://"
    var hdfsMasterPath = hdfsMasterBase + "/user/oas/"
    var ftpIP = ""
    var ftpUser = "root"
    var ftpPW = "123456"
    /*hdfs目录*/
    var hdfsWorkPath = "/home/"
    /*ftp目录*/
    var ftpBasePath = "/"
    /*hdfs用户名*/
    var hdfsUserName = "oas"
    /*ftp文件reg*/
    var regExp = "^[A-Za-z]{3}_(SEGMENT|CLASS){1}_\\d{8}\\.csv$"
    /*ftp目录搜索深度*/
    var dept = 7
    /*保存下载记录前10000条的文件路径*/
    var downloadedLogFilePath = "/user/oas/downloaded_ftp_files.txt"
    var transLower = false
    /*保存下载记录前10000条的set*/
    var downloadedFiles = new mutable.HashSet[(String, Long)]()

    /*保存下载记录前10000条的set*/
    def main(args: Array[String]): Unit = {
        if (args != null) {
            println("args.length: " + args.length)
            for (s <- args) {
                println("args: " + s)
            }
        }
        if (args != null && args.length > 4) {
            ftpBasePath = args(0)
            hdfsWorkPath = args(1)
            hdfsUserName = args(2)
            regExp = args(3)
            dept = Integer.valueOf(args(4))
            if(args.length > 5) transLower = args(5).toBoolean
        }

        val conf = new SparkConf().setAppName("ftp files download")
                .set("spark.sql.parquet.compression.codec","uncompressed")
        //gzip
        val spark = new SparkContext(conf)
        System.setProperty("HADOOP_USER_NAME",hdfsUserName)
        try{
            //先检查主备状态
            val validUri = GlobalInstance.checkHdfsUri()
            if (validUri.length >0){
                hdfsMasterBase = validUri(0)
                println("----start hdfs"+hdfsMasterBase)
                downloadedFiles = readDownloadLogFile
                ftpFilesDownload(hdfsMasterPath,ftpBasePath,regExp,dept)
            }else{
                println("---no valid hdfs")
            }
        }catch {
            case e:Exception => e.printStackTrace()
        }
        spark.stop()
    }

    private def readDownloadLogFile:mutable.HashSet[(String,Long)]={
        val set = new mutable.HashSet[(String,Long)]()
        try{

        }catch {
            case e:Exception=>e.printStackTrace()
        }
        set
    }

    /*ftp文件下载*/
    private def ftpFilesDownload(
                                hdfsPath:String,
                                ftpPath:String,
                                regExp:String,
                                dept:Int
                                )={
        val files = new ArrayBuffer[(String,Long)](5000)
        var ftp = ftpReconnect()
        ftpListFile(ftp,ftpPath,regExp,files,dept)
        ftp.changeWorkingDirectory(ftpPath)
        files.foreach{
            ftpFile =>
                try{
                    val file = ftpFile._1
                    val separatorIndex = file.lastIndexOf("/")
                    val fileName = file.substring(separatorIndex+1)
                    val filePath = file.substring(0,separatorIndex)
                    ftp = ftpReconnect()
                    ftp.changeWorkingDirectory(ftpPath)
                    val inSteam = ftp.retrieveFileStream(fileName)
                    val fileSystem =GlobalInstance.getHdfsFileSystem(hdfsMasterBase,hdfsUserName)
                    val outStream = fileSystem
                            .create(new Path(hdfsPath + (if(transLower) fileName.toLowerCase else fileName)), true) //文件名存为小写
                    IOUtils.copyBytes(inSteam,outStream,81920)
                    outStream.close()
                    inSteam.close()
                    ftp.disconnect()
                    println(file+"dowload")
                    downloadedFiles+=ftpFile
                }catch {
                    case e:Exception=>e.printStackTrace()
                }
        }

        //复写前10000个下载记录
        try{
            val fileSystem = GlobalInstance.getHdfsFileSystem(hdfsMasterBase,hdfsUserName)
            val outputStream = fileSystem.create(new Path(downloadedLogFilePath),true)
            val bufferedWriter = new BufferedWriter(new OutputStreamWriter(outputStream))
            val result = downloadedFiles.toArray.sortWith((a,b)=>a._2>b._2)
            //记录最近下载的10000
            val size =if(result.length > 10000)10000 else result.length
            for ( i <- 0 until size){
                bufferedWriter.write(result.apply(i).toString())
                bufferedWriter.newLine()
            }
            bufferedWriter.close()
        }catch {
            case e:Exception=>e.printStackTrace()
        }
    }

    def ftpListFile(ftpClient:FTPClient,
                    ftpPath:String,
                    regExp:String,
                    files:ArrayBuffer[(String,Long)],
                    dept:Int): Unit ={
        var anotherDept = dept//复制基本类型int dept
        if(anotherDept <= 0){
            return
        }
        var ftp = ftpClient
        if(!ftp.isConnected){
            ftp = ftpReconnect()
        }
        val directory = ftpPath
        //更换目录到当前目录
        ftp.changeWorkingDirectory(directory)
        val ftpFiles = ftp.listFiles()
        anotherDept -= 1
        for (file <- ftpFiles){
            if(file.isFile){
                if(file.getName.matches(regExp)){
                    val one = (directory+file.getName,file.getTimestamp.getTimeInMillis)
                    if(!downloadedFiles.contains(one)){
                        files += one
                        println("download file:"+directory+file.getName+",size"+file.getSize)
                    }else{
                        println("ignore file"+directory+file.getName+"size"+file.getSize)
                    }
                }else{
                    ftpListFile(ftp,directory+file.getName+"/",regExp,files,anotherDept)
                }
            }
        }
    }

    def ftpReconnect():FTPClient = {
        val ftp = new FTPClient()
        ftp.setControlEncoding("UTF-8")
        ftp.setDefaultTimeout(72000)
        ftp.connect(ftpIP,21)
        ftp.enterLocalPassiveMode()
        ftp.login(ftpUser,ftpPW)
        ftp.setFileType(FTP.BINARY_FILE_TYPE)
        ftp
    }








}
